libp2p/autotls/peeridauth.nim

```diff
@@ -0,0 +1,229 @@
+  k: string
+  v: seq[byte]
+
+proc randomChallenge(
+    rng: ref HmacDrbgContext, challengeLen: int = ChallengeDefaultLen
+): string =
+  var rng = rng[]
+  var challenge = ""
+  for _ in 0 ..< challengeLen:
+    challenge.add(rng.sampleChar(ChallengeCharset))
+  challenge
+
+proc extractField(data, key: string): string {.raises: [PeerIDAuthError].} =
+  # Helper to extract quoted value from key
+  for segment in data.split(","):
+    if key in segment:
+      return segment.split("=", 1)[1].strip(chars = {' ', '"'})
+  raise newException(PeerIDAuthError, "Failed to find " & key & " in " & data)
+
+proc encodeVarint(n: int): seq[byte] =
+  var varInt: seq[byte] = @[]
+  var x = uint64(n)
+  while x != 0:
+    var byteVal = byte(x and 0x7F)
+    x = x shr 7
+    if x != 0:
+      byteVal = byteVal or 0x80
+    varInt.add(byteVal)
+  return varInt
+
+proc genDataToSign(prefix: string, parts: seq[SigParam]): seq[byte] =
+  var buf: seq[byte] = prefix.toByteSeq()
+  for p in parts:
+    buf.add encodeVarint(p.k.len + p.v.len + 1)
+    buf.add (p.k & "=").toByteSeq()
+    buf.add p.v
+  return buf
+
+proc getSigParams(
+    clientSender: bool, hostname: string, challenge: string, publicKey: PublicKey
+): seq[SigParam] =
+  if clientSender:
+    @[
+      SigParam(k: "challenge-client", v: challenge.toByteSeq()),
+      SigParam(k: "hostname", v: hostname.toByteSeq()),
+      SigParam(k: "server-public-key", v: publicKey.getBytes().get()),
+    ]
+  else:
+    @[
+      SigParam(k: "challenge-server", v: challenge.toByteSeq()),
+      SigParam(k: "client-public-key", v: publicKey.getBytes().get()),
+      SigParam(k: "hostname", v: hostname.toByteSeq()),
+    ]
+
+proc peerIdSign(
+    privateKey: PrivateKey,
+    challenge: string,
+    publicKey: PublicKey,
+    hostname: string,
+    clientSender: bool = true,
+): string =
+  let parts = getSigParams(clientSender, hostname, challenge, publicKey)
+  let bytesToSign = genDataToSign(PeerIDAuthPrefix, parts)
+  return base64.encode(privateKey.sign(bytesToSign).get().getBytes(), safe = true)
+
+proc checkSignature(
+    serverSig: string,
+    serverPublicKey: PublicKey,
+    challengeServer: string,
+    clientPublicKey: PublicKey,
+    hostname: string,
+): bool {.raises: [PeerIDAuthError].} =
+  let parts = getSigParams(false, hostname, challengeServer, clientPublicKey)
+  let signedBytes = genDataToSign(PeerIDAuthPrefix, parts)
+  var serverSignature: Signature
+  try:
+    if not serverSignature.init(base64.decode(serverSig).toByteSeq()):
+      raise newException(
+        PeerIDAuthError, "Failed to initialize Signature from base64 encoded sig"
+      )
+  except ValueError as exc:
+    raise newException(PeerIDAuthError, "Failed to decode server's signature", exc)
+  return serverSignature.verify(
+    signedBytes.toOpenArray(0, signedBytes.len - 1), serverPublicKey
+  )
+
+proc peerIdAuthenticate(
+    url: string,
+    session: HttpSessionRef,
+    peerInfo: PeerInfo,
+    payload: JsonNode,
```

Comment:
should not use `JsonNode` as payload, pass exactly values (with types) that you need, and create payload inside.

Reply:
Peeridauth is an authentication scheme. At the end of the exchange the client will send data to the server. What data that client will send can greatly vary as this is not an autotls-specific protocol, autotls just uses it to authenticate node and broker. This function should not handle parsing. Parsing should be handled outside of it